client/src/app/components/Toolbar/Toolbar.jsx

```diff
@@ -19,27 +19,65 @@ import { createShoppingList } from 'modules/shopping-list/model/actions';
 import { createCohort } from 'modules/cohort/model/actions';
 import { getCurrentUser } from 'modules/authorization/model/selectors';
 import { UserPropType } from 'common/constants/propTypes';
+import Overlay, { OverlayStyleType } from 'common/components/Overlay';
 
 class Toolbar extends PureComponent {
   state = {
-    cohortFormVisbility: false,
-    shoppingFormVisbility: false
+    cohortFormVisibility: false,
+    overlayVisibility: false,
+    shoppingFormVisibility: false,
+    userBarMenuVisibility: false
+  };
+
+  componentDidMount() {
+    document.addEventListener('click', this.clickListener);
+    document.addEventListener('keydown', this.escapeListener);
+  }
+
+  componentWillUnmount() {
+    document.removeEventListener('click', this.clickListener);
+    document.removeEventListener('keydown', this.escapeListener);
+  }
+
+  clickListener = event => {
+    const { className } = event.target;
+    className.length > 0 && className.includes('overlay')
+      ? this.hideOverlayAndForm()
+      : null;
+  };
+
+  escapeListener = event => {
+    event.code === 'Escape' ? this.hideOverlayAndForm() : null;
+  };
+
+  hideOverlayAndForm = () => {
+    this.setState({
+      cohortFormVisibility: false,
+      overlayVisibility: false,
+      shoppingFormVisibility: false,
+      userBarMenuVisibility: false
+    });
   };
 
   handleShoppingListFormVisibility = () => {
-    const { shoppingFormVisbility } = this.state;
+    const { shoppingFormVisibility } = this.state;
+
     this.setState({
-      shoppingFormVisbility: !shoppingFormVisbility,
-      cohortFormVisbility: false
+      cohortFormVisibility: false,
+      overlayVisibility: !shoppingFormVisibility,
+      shoppingFormVisibility: !shoppingFormVisibility,
+      userBarMenuVisibility: false
```

Comment:
only wonder if this is scalable solution: If the button has its own handler then if we add more buttons we will have of course more handlers, and in each of them all the props  to change... If you remove button you have to edit all handlers...

Reply:
every form handles independent part of api. any idea how to handle this form visibility and submission in another more scalable way?

Reply:
I wonder if we could use (To avoid using overlay):
- additional option in addeventlistener: capture  -  and set it to true
- detect if clicked element is a child of form to not close the form